doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,381 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _codegen:
+
+(Inline) Code Generation
+########################
+
+.. contents::
+   :depth: 2
+   :local:
+   :backlinks: top
+
+Description
+***********
+
+Python snippets that are inlined in a source file are used as code generators.
+The tool to scan the source file for the Python snippets and process them is
+Codegen. Codegen and part of this documentation is based on
+`Cog <https://nedbatchelder.com/code/cog/index.html>`_ from Ned Batchelder.
+
+The processing of source files is controlled by two CMake extension functions:
+zephyr_sources_codegen(..) and zephyr_sources_codegen_ifdef(..). During CMake
+configuration the source files are processed by Codegen and the generated
+source files are written to the CMake binary directory. The generated source
+files are added to the Zephyr sources.
+
+The inlined Python snippets can contain any Python code, they are regular
+Python scripts. All Python snippets in a source file and all Python snippets of
+included template files are treated as a python script with a common set of
+global Python variables. Global data created in one snippet can be used in
+another snippet that is processed later on. This feature is e.g. used to
+customize included template files.
+
+An inlined Python snippet can always access the codegen module. The codegen
+module encapsulates and provides all the functions to retrieve information
+(options, device tree properties, CMake variables, config properties) and to
+put out the generated code.
+
+Codegen transforms files in a very simple way: it finds chunks of Python code
+embedded in them, executes the Python code, and inserts its output back into
+the original file. The file can contain whatever text you like around the
+Python code. It will usually be source code.
+
+For example, if you run this file through Codegen:
+
+::
+
+    /* This is my C file. */
+    ...
+    /**
```

Comment:
This does not conflict with doxygen. It this example it is by purpose to include the python code snippet into doxygen output. The markers for code @code{.codegen}, .. are chosen in a way to support documentation by doxygen. /** is a marker for doxygen only and can be replaced by /* to leave the code snippet out of doxygen output..

Reply:
OK.  I'm curious how this will play with the document generating process, and if we'll need to add a codegen pass over source files before presenting them to doxygen.

Reply:
> how this will play with the document generating process

Currently none of the files are parsed by doxygen because the driver directory is not included in the doxygen documentation.

>  if we'll need to add a codegen pass over source

There is no need for an extra pass. Inline code is within comments clearly marked with the doxygen special commands @code/@endcode. Doxygen does not know the {.codegen} type.  It ignores the unknown code type specification (see https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcode) and "just show the output as-is". 

As an improvement {.codegen} could be made known to doxygen to be treated as python code.

Reply:
Thanks for clarifying!